File: lib/android/app/src/main/java/com/reactnativenavigation/options/parsers/IconParser.java
package com.reactnativenavigation.options.parsers;

import android.content.Context;

import com.reactnativenavigation.options.params.NullText;
import com.reactnativenavigation.options.params.Text;

import org.json.JSONException;
import org.json.JSONObject;

import javax.annotation.Nullable;

public class IconParser {
    private static Integer getSystemResourceId(String systemIcon, Context context) {
        return context.getResources().getIdentifier(systemIcon, "drawable", "android");
    }

    public static Text parse(@Nullable JSONObject json, String key, Context context) {
        if (json == null || !json.has(key)) return new NullText();
        try {
            if (!TextParser.parse(json.optJSONObject(key), "system").hasValue()) {
                Text fallbackIcon = TextParser.parse(json.optJSONObject(key), "fallback");
                Text systemIcon = TextParser.parse(json.optJSONObject(key), "system");
                if (getSystemResourceId(systemIcon.toString(), context) > 0) {
                    return new Text("@android:drawable/" + systemIcon);
                } else {
                    if (!fallbackIcon.hasValue()) {
                        JSONObject js = json.optJSONObject(key);
                        if (js != null) {
                            return js.get("fallback") instanceof String ? TextParser.parse(js, "fallback") : TextParser.parse(js.optJSONObject("fallback"), "uri");
                        } else {
                            return new NullText();
                        }
                    }
                }
            }
            return json.get(key) instanceof String ? TextParser.parse(json, key) : TextParser.parse(json.optJSONObject(key), "uri");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return new NullText();
    }
}

File: lib/android/app/src/main/java/com/reactnativenavigation/options/parsers/IconParser.kt
package com.reactnativenavigation.options.parsers

import com.reactnativenavigation.options.params.NullTextProp
import com.reactnativenavigation.options.params.TextProp
import org.json.JSONException
import org.json.JSONObject

object IconParser {
    @JvmStatic
    fun parse(json: JSONObject?, key: String): TextProp {
        if (json == null || !json.has(key)) return NullTextProp
        try {
            return if (json[key] is String) TextParser.parse(json, key) else TextParser.parse(json.optJSONObject(key), "uri")
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return NullTextProp
    }
}